birdhouse/optional-components/canarie-api-full-monitoring/default.env

```diff
@@ -2,5 +2,6 @@
 
 # add any component that this component requires to run
 COMPONENT_DEPENDENCIES="
+  ./optional-components/all-public-access
```

Comment:
Should not be here.  Not all org want to make everything public so activating all-public-access component should be explicit by each org and not implicit like this.

This brings another question.  Misha all-public-access will grant public access only if the WPS component is enabled.  Does this canarie-api-full-monitoring having handling for not enabled components?

Reply:
When activating `canarie-api-full-monitoring`, minimal permission required would be `GetCapabilities` for WPS-based services, and `read-match` on the top-most resource for RESTful-based services.

Those permissions could be set directly in the `canarie-api-full-monitoring` definition, so the access requirements are activated at the same time. However, it would be important that those permissions be loaded first (load order of Magpie permission config files) to allow broader `read` (if set anywhere) to override the `read-match`. 

Regarding the handling, the configs use a loop over injected birds in:
`birdhouse/optional-components/canarie-api-full-monitoring/config/canarie-api/canarie_api_full_monitoring.py.template`
Those bird configs are injected according to enabled components. So permissions should only be applied for enabled services.

Reply:
Are you proposing for the `canarie-api-full-monitoring` to handle its own read permissions?

I prefer to leave it like this to avoid surprise for the user (user don't know where the permission are coming from and do not expect it to be from `canarie-api-full-monitoring`).

It is an optional components after all.  It should not start fiddling with Magpie permissions.

Reply:
What if we set the required permissions for canarie-api-full-monitoring in `canarie-api-full-monitoring/config/magpie` so that it is clear that these permissions are required in order to make it work (without having to rely on all-public-access)

Reply:
> Are you proposing for the canarie-api-full-monitoring to handle its own read permissions?

I think that would be the most logical approach. Avoid extra steps from users.
The `canarie-api-full-monitoring` already states this, so there shouldn't be a surprise: 

https://github.com/bird-house/birdhouse-deploy/tree/master/birdhouse/optional-components#monitor-all-components-in-canarie-node-both-public-and-internal-url
> This assume all the WPS services are public. If not the case, make a copy of this config and adjust accordingly.

I think users would be even more surprised to see errors of something they just enabled (it should work out-of-the-box), and asking them to correctly understand the Magpie permission configuration to correct accesses, which require understanding each service type/permission scheme, makes it much more complicated for them.

Reply:
I think I need to explain why that `canarie-api-full-monitoring` component came to be.

We all agree that ensuring the public access works is a good idea.  So why this component is optional and there is a notice saying "If not the case, make a copy of this config and adjust accordingly".  Probably that notice is not descriptive enough, my bad.

It was chosen to put all the WPS birds behind Twitcher for access control.  Furthermore, it is a very fine grained access control.  Each specific WPS action (GetCapabilities, DescribeProcess, ...) can be toggle independently for each WPS service.

I was not there at the time it was designed that way, but I assume we want a strict control over which WPS service and capabilities we want to allow.  So maybe an org can deploy all the birds but only want to allow complete public access for some only, but not all of them. 

So that's why this very useful `canarie-api-full-monitoring` component is optional and not integrated into to main `canarie-api`.  The notice "If not the case, make a copy of this config and adjust accordingly" means make a copy of that config and remove the birds you do not want to make public.

At Ouranos, all the birds are public.  But if we take PCIC for example, I know they do not use any of the birds.  They have their own birds they are plugin in via external repos.  So they can make a copy of this config, re-adjust for their birds, and deploy together with their birds via their external repo.  They do not need to activate this `canarie-api-full-monitoring` as-is.

Furthermore, Magpie permissions once given via a component, is not revoked by simply disabling that component.  You have to go to Magpie UI to manually remove the access.  So if someone want just to try out this component, then decide they do not want it, all their birds will still be wide open to the internet, even after they disabled the component.

The `all-public-access` component https://github.com/bird-house/birdhouse-deploy/tree/master/birdhouse/optional-components#give-public-access-to-all-resources-for-testing-purposes state that very clearly "Once enabled, if you need to revert the change, you have to do it manually by logging into Magpie. Just disabling this component will not revert the change."

So putting our sysadmin hat, we really do not want a silent surprise that all the birds are public on the internet without us explicitly knowingly allow it.

So that's why I much prefer this `canarie-api-full-monitoring` component stay a bit "dumb" and relies on either manual explicit intervention on Magpie WebUI or explicit activation of `all-public-access` if on a staging or test system.  Either ways, the sysadmin knowingly did it.  It's not a surprise.

I understand this is not the "it just works out of the box" experience we are looking for, but otherwise it would be a false sense of security to put Twitcher there but then activate public access everywhere and the sysadmin has to manually remember to go disable the accesses they do not want.

So instead of this notice for `canarie-api-full-monitoring` 
  "This assume all the WPS services are public. If not the case, make a copy of this config and adjust accordingly", 
would this be better 
  "This assume all the WPS services are public. If not the case, because not all WPS services are required to be publicly accessible or additional custom WPS services are enabled, you can make a copy of this config, or even this entire component, and add or remove WPS services according to the services that are public only."?

Reply:
Please remove this dependency on `all-public-access` component because this is too sneaky on the users.  They do not realized they are opening up to the wide internet and the only way out is manual edit in Magpie UI.

Reply:
>Magpie permissions once given via a component, is not revoked by simply disabling that component. You have to go to Magpie UI to manually remove the access. So if someone want just to try out this component, then decide they do not want it, all their birds will still be wide open to the internet, even after they disabled the component.

This is true. 
The right way to handle this automatically would be to have a component that lists all known permissions for "opening" services publicly, and use the `remove` action instead. 

I agree that `all-public-access` is too open to be used as dependency, this is why `canarie-api-full-monitoring` should set only the permissions it needs (basically just the `read` or `GetCapabilities` of the top most monitored resource.

Reply:
@mishaschwartz 
> - create a dedicated user for canarie-api in magpie
> - allow that user limited access permissions to the routes we want to monitor
> - canarie-api then can access these routes as that user
>
> I know you mentioned in the past that you don't love the idea of having internal users registered in magpie [...]

That would work. I don't recall mentioning that internal users are bad, as I also think the approach you propose is an ideal way to handle this issue. Some considerations though. 
All requests in CanarieAPI will have to be updated to use `requests-magpie` auth hook to perform the login/cookie session handling if credentials are provided. Also, `proxy` (or eventually `canarieapi` if separated out) will need to depend on Magpie to ensure it is ready to register users and perform the login.

Reply:
@tlvu @mishaschwartz 
For now, I reverted the `all-public-access` without any extra permission handling. 
On the CI side, `all-public-access` is added explicitly, so everything should still work for PR tests.
Users will have to add custom permissions in the meantime if they want to provide access to CanarieAPI.